File: packages/agent-cache/src/tiers/ToolCache.ts
import type { Valkey, ToolStoreOptions, ToolCacheResult, ToolPolicy } from '../types';
import type { Telemetry } from '../telemetry';
import { ValkeyCommandError, AgentCacheUsageError } from '../errors';
import { toolCacheHash, escapeGlobPattern } from '../utils';
import { clusterScan } from '../cluster';

/**
 * Validate that tool name doesn't contain colons, which are used as key delimiters.
 * Tool names with colons would break stats parsing in AgentCache.stats().
 */
function validateToolName(toolName: string): void {
  if (toolName.includes(':')) {
    throw new AgentCacheUsageError(
      `Tool name "${toolName}" contains colon (:). ` +
      `Colons are not allowed in tool names as they are used as key delimiters.`
    );
  }
}

export interface ToolCacheConfig {
  client: Valkey;
  name: string;
  defaultTtl: number | undefined;
  tierTtl: number | undefined;
  telemetry: Telemetry;
  statsKey: string;
}

interface StoredToolEntry {
  response: string;
  toolName: string;
  args: unknown;
  storedAt: number;
  cost?: number;
}

export class ToolCache {
  private readonly client: Valkey;
  private readonly name: string;
  private readonly defaultTtl: number | undefined;
  private readonly tierTtl: number | undefined;
  private readonly telemetry: Telemetry;
  private readonly statsKey: string;
  private readonly policies: Map<string, ToolPolicy> = new Map();
  private readonly policiesKey: string;

  constructor(config: ToolCacheConfig) {
    this.client = config.client;
    this.name = config.name;
    this.defaultTtl = config.defaultTtl;
    this.tierTtl = config.tierTtl;
    this.telemetry = config.telemetry;
    this.statsKey = config.statsKey;
    this.policiesKey = `${this.name}:__tool_policies`;
  }

  private buildKey(toolName: string, hash: string): string {
    return `${this.name}:tool:${toolName}:${hash}`;
  }

  async check(toolName: string, args: unknown): Promise<ToolCacheResult> {
    validateToolName(toolName);
    const startTime = Date.now();

    return this.telemetry.tracer.startActiveSpan('agent_cache.tool.check', async (span) => {
      try {
        const hash = toolCacheHash(args);
        const key = this.buildKey(toolName, hash);

        span.setAttribute('cache.key', key);
        span.setAttribute('cache.tool_name', toolName);

        let raw: string | null;
        try {
          raw = await this.client.get(key);
        } catch (err) {
          throw new ValkeyCommandError('GET', err);
        }

        const duration = (Date.now() - startTime) / 1000;
        this.telemetry.metrics.operationDuration
          .labels(this.name, 'tool', 'check')
          .observe(duration);

        if (raw) {
          let entry: StoredToolEntry;
          try {
            entry = JSON.parse(raw);
          } catch {
            // Corrupt cache entry - await delete to guarantee cleanup before returning miss
            await this.client.del(key).catch(() => {});
            try {
              const statsPipeline = this.client.pipeline();
              statsPipeline.hincrby(this.statsKey, 'tool:misses', 1);
              statsPipeline.hincrby(this.statsKey, `tool:${toolName}:misses`, 1);
              await statsPipeline.exec();
            } catch {
              // Stats update failure should not break the cache
            }
            this.telemetry.metrics.requestsTotal
              .labels(this.name, 'tool', 'miss', toolName)
              .inc();
            span.setAttribute('cache.hit', false);
            span.setAttribute('cache.corrupt', true);
            span.end();
            return { hit: false, tier: 'tool' as const, toolName };
          }

          // Record tier-level and per-tool hit + cost savings in a single pipeline
          try {
            const statsPipeline = this.client.pipeline();
            statsPipeline.hincrby(this.statsKey, 'tool:hits', 1);
            statsPipeline.hincrby(this.statsKey, `tool:${toolName}:hits`, 1);
            if (entry.cost !== undefined) {
              const costMicros = Math.round(entry.cost * 1_000_000);
              statsPipeline.hincrby(this.statsKey, 'cost_saved_micros', costMicros);
              statsPipeline.hincrby(this.statsKey, `tool:${toolName}:cost_saved_micros`, costMicros);
            }
            await statsPipeline.exec();
          } catch {
            // Stats update failure should not break the cache
          }

          // Track cost in Prometheus (outside pipeline since it's local)
          if (entry.cost !== undefined) {
            this.telemetry.metrics.costSaved
              .labels(this.name, 'tool', '', toolName)
              .inc(entry.cost);
          }

          this.telemetry.metrics.requestsTotal
            .labels(this.name, 'tool', 'hit', toolName)
            .inc();

          span.setAttribute('cache.hit', true);
          span.end();

          return {
            hit: true,
            response: entry.response,
            key,
            tier: 'tool' as const,
            toolName,
          };
        }

        // Record tier-level and per-tool miss (batch into pipeline)
        try {
          const statsPipeline = this.client.pipeline();
          statsPipeline.hincrby(this.statsKey, 'tool:misses', 1);
          statsPipeline.hincrby(this.statsKey, `tool:${toolName}:misses`, 1);
          await statsPipeline.exec();
        } catch {
          // Stats update failure should not break the cache
        }

        this.telemetry.metrics.requestsTotal
          .labels(this.name, 'tool', 'miss', toolName)
          .inc();

        span.setAttribute('cache.hit', false);
        span.end();

        return {
          hit: false,
          tier: 'tool' as const,
          toolName,
        };
      } catch (err) {
        span.recordException(err as Error);
        span.end();
        throw err;
      }
    });
  }

  async store(toolName: string, args: unknown, response: string, options?: ToolStoreOptions): Promise<string> {
    validateToolName(toolName);
    const startTime = Date.now();

    return this.telemetry.tracer.startActiveSpan('agent_cache.tool.store', async (span) => {
      try {
        const hash = toolCacheHash(args);
        const key = this.buildKey(toolName, hash);

        span.setAttribute('cache.key', key);
        span.setAttribute('cache.tool_name', toolName);

        // Store cost in entry - cost tracking happens at check() time on hit, not here
        const entry: StoredToolEntry = {
          response,
          toolName,
          args,
          storedAt: Date.now(),
          cost: options?.cost,
        };

        const valueJson = JSON.stringify(entry);

        // TTL resolution order: per-call -> policy -> tier -> default
        const policy = this.policies.get(toolName);
        const ttl = options?.ttl ?? policy?.ttl ?? this.tierTtl ?? this.defaultTtl;

        // Use SET with EX option for atomic set+expire to prevent orphaned keys
        try {
          if (ttl !== undefined) {
            await this.client.set(key, valueJson, 'EX', ttl);
          } else {
            await this.client.set(key, valueJson);
          }
        } catch (err) {
          throw new ValkeyCommandError('SET', err);
        }

        // Track stored bytes (measure valueJson, not just response, since that's what's stored)
        const byteLength = Buffer.byteLength(valueJson, 'utf8');
        this.telemetry.metrics.storedBytes
          .labels(this.name, 'tool')
          .inc(byteLength);

        const duration = (Date.now() - startTime) / 1000;
        this.telemetry.metrics.operationDuration
          .labels(this.name, 'tool', 'store')
          .observe(duration);

        span.setAttribute('cache.ttl', ttl ?? -1);
        span.setAttribute('cache.bytes', byteLength);
        span.end();

        return key;
      } catch (err) {
        span.recordException(err as Error);
        span.end();
        throw err;
      }
    });
  }

  async setPolicy(toolName: string, policy: ToolPolicy): Promise<void> {
    validateToolName(toolName);
    this.policies.set(toolName, policy);

    // Persist to Valkey
    try {
      await this.client.hset(this.policiesKey, toolName, JSON.stringify(policy));
    } catch (err) {
      throw new ValkeyCommandError('HSET', err);
    }
  }

  getPolicy(toolName: string): ToolPolicy | undefined {
    return this.policies.get(toolName);
  }

  /** Returns the names of tools with persisted policies. Used by the discovery marker. */
  listPolicyNames(): string[] {
    return Array.from(this.policies.keys());
  }

  async invalidateByTool(toolName: string): Promise<number> {
    return this.telemetry.tracer.startActiveSpan('agent_cache.tool.invalidateByTool', async (span) => {
      try {
        span.setAttribute('cache.tool_name', toolName);

        // Escape glob chars to match only this tool's keys during SCAN.
        const pattern = `${escapeGlobPattern(this.name)}:tool:${escapeGlobPattern(toolName)}:*`;
        let deletedCount = 0;

        await clusterScan(this.client, pattern, async (keys, nodeClient) => {
          // Pipeline DEL — individual commands avoid CROSSSLOT in cluster mode
          const pipeline = nodeClient.pipeline();
          for (const key of keys) pipeline.del(key);
          let delResults: Array<[Error | null, number]>;
          try {
            delResults = await pipeline.exec() as Array<[Error | null, number]>;
          } catch (err) {
            throw new ValkeyCommandError('DEL', err);
          }
          for (const [err, count] of delResults) {
            if (err) throw new ValkeyCommandError('DEL', err);
            deletedCount += count ?? 0;
          }
        });

        span.setAttribute('cache.deleted_count', deletedCount);
        span.end();

        return deletedCount;
      } catch (err) {
        span.recordException(err as Error);
        span.end();
        throw err;
      }
    });
  }

  async invalidate(toolName: string, args: unknown): Promise<boolean> {
    const hash = toolCacheHash(args);
    const key = this.buildKey(toolName, hash);

    try {
      const deleted = await this.client.del(key);
      return deleted > 0;
    } catch (err) {
      throw new ValkeyCommandError('DEL', err);
    }
  }

  async loadPolicies(): Promise<void> {
    try {
      const raw = await this.client.hgetall(this.policiesKey);
      if (raw) {
        for (const [toolName, policyJson] of Object.entries(raw)) {
          try {
            const policy: ToolPolicy = JSON.parse(policyJson);
            this.policies.set(toolName, policy);
          } catch {
            // Skip corrupt policy entries
          }
        }
      }
    } catch {
      // Non-blocking: failure to load policies should not break initialization.
      // Silently swallow - libraries should not write to console.
    }
  }

  /**
   * Clear in-memory tool policies. Called by AgentCache.flush() to stay in sync
   * after all Valkey keys (including __tool_policies) are deleted.
   */
  resetPolicies(): void {
    this.policies.clear();
  }
}

File: packages/agent-cache-py/betterdb_agent_cache/tiers/tool_cache.py
from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from ..cluster import cluster_scan
from ..errors import AgentCacheUsageError, ValkeyCommandError
from ..types import ToolCacheResult, ToolPolicy, ToolStoreOptions
from ..utils import escape_glob_pattern, tool_cache_hash

if TYPE_CHECKING:
    from ..telemetry import Telemetry


def _validate_tool_name(tool_name: str) -> None:
    if ":" in tool_name:
        raise AgentCacheUsageError(
            f'Tool name "{tool_name}" contains colon (:). '
            "Colons are not allowed in tool names as they are used as key delimiters."
        )


@dataclass
class ToolCacheConfig:
    client: Any
    name: str
    default_ttl: int | None
    tier_ttl: int | None
    telemetry: Telemetry
    stats_key: str


class ToolCache:
    def __init__(self, config: ToolCacheConfig) -> None:
        self._client = config.client
        self._name = config.name
        self._default_ttl = config.default_ttl
        self._tier_ttl = config.tier_ttl
        self._telemetry = config.telemetry
        self._stats_key = config.stats_key
        self._policies: dict[str, ToolPolicy] = {}
        self._policies_key = f"{self._name}:__tool_policies"

    def _build_key(self, tool_name: str, hash_: str) -> str:
        return f"{self._name}:tool:{tool_name}:{hash_}"

    async def check(self, tool_name: str, args: Any) -> ToolCacheResult:
        _validate_tool_name(tool_name)
        start = time.monotonic()
        with self._telemetry.tracer.start_as_current_span("agent_cache.tool.check") as span:
            try:
                hash_ = tool_cache_hash(args)
                key = self._build_key(tool_name, hash_)
                span.set_attribute("cache.key", key)
                span.set_attribute("cache.tool_name", tool_name)

                try:
                    raw = await self._client.get(key)
                except Exception as exc:
                    raise ValkeyCommandError("GET", exc) from exc

                duration = time.monotonic() - start
                self._telemetry.metrics.operation_duration.labels(
                    self._name, "tool", "check"
                ).observe(duration)

                if raw:
                    raw_str = raw.decode() if isinstance(raw, bytes) else raw
                    try:
                        entry: dict[str, Any] = json.loads(raw_str)
                    except (json.JSONDecodeError, ValueError):
                        try:
                            await self._client.delete(key)
                        except Exception:
                            pass
                        await self._inc_stats({
                            "tool:misses": 1,
                            f"tool:{tool_name}:misses": 1,
                        })
                        self._telemetry.metrics.requests_total.labels(
                            self._name, "tool", "miss", tool_name
                        ).inc()
                        span.set_attribute("cache.hit", False)
                        span.set_attribute("cache.corrupt", True)
                        return ToolCacheResult(hit=False, tool_name=tool_name)

                    stat_updates: dict[str, int] = {
                        "tool:hits": 1,
                        f"tool:{tool_name}:hits": 1,
                    }
                    cost: float | None = entry.get("cost")
                    if cost is not None:
                        cost_micros = round(cost * 1_000_000)
                        stat_updates["cost_saved_micros"] = cost_micros
                        stat_updates[f"tool:{tool_name}:cost_saved_micros"] = cost_micros
                    await self._inc_stats(stat_updates)

                    if cost is not None:
                        self._telemetry.metrics.cost_saved.labels(
                            self._name, "tool", "", tool_name
                        ).inc(cost)

                    self._telemetry.metrics.requests_total.labels(
                        self._name, "tool", "hit", tool_name
                    ).inc()
                    span.set_attribute("cache.hit", True)
                    return ToolCacheResult(
                        hit=True,
                        tool_name=tool_name,
                        response=entry.get("response"),
                        key=key,
                    )

                await self._inc_stats({
                    "tool:misses": 1,
                    f"tool:{tool_name}:misses": 1,
                })
                self._telemetry.metrics.requests_total.labels(
                    self._name, "tool", "miss", tool_name
                ).inc()
                span.set_attribute("cache.hit", False)
                return ToolCacheResult(hit=False, tool_name=tool_name)

            except Exception as exc:
                span.record_exception(exc)
                raise

    async def store(
        self,
        tool_name: str,
        args: Any,
        response: str,
        options: ToolStoreOptions | None = None,
    ) -> str:
        _validate_tool_name(tool_name)
        start = time.monotonic()
        with self._telemetry.tracer.start_as_current_span("agent_cache.tool.store") as span:
            try:
                hash_ = tool_cache_hash(args)
                key = self._build_key(tool_name, hash_)
                span.set_attribute("cache.key", key)
                span.set_attribute("cache.tool_name", tool_name)

                entry: dict[str, Any] = {
                    "response": response,
                    "toolName": tool_name,
                    "args": args,
                    "storedAt": int(time.time() * 1000),
                }
                if options and options.cost is not None:
                    entry["cost"] = options.cost

                value_json = json.dumps(entry, separators=(",", ":"), ensure_ascii=False)

                # TTL resolution: per-call → policy → tier → default
                policy = self._policies.get(tool_name)
                _per_call_ttl = options.ttl if options else None
                _policy_ttl = policy.ttl if policy else None
                ttl = next(
                    (t for t in (_per_call_ttl, _policy_ttl, self._tier_ttl, self._default_ttl) if t is not None),
                    None,
                )
                await self._set(key, value_json, ttl)

                byte_len = len(value_json.encode())
                self._telemetry.metrics.stored_bytes.labels(self._name, "tool").inc(byte_len)
                self._telemetry.metrics.operation_duration.labels(
                    self._name, "tool", "store"
                ).observe(time.monotonic() - start)

                span.set_attribute("cache.ttl", ttl if ttl is not None else -1)
                span.set_attribute("cache.bytes", byte_len)
                return key

            except Exception as exc:
                span.record_exception(exc)
                raise

    async def set_policy(self, tool_name: str, policy: ToolPolicy) -> None:
        _validate_tool_name(tool_name)
        self._policies[tool_name] = policy
        try:
            await self._client.hset(
                self._policies_key, tool_name, json.dumps({"ttl": policy.ttl})
            )
        except Exception as exc:
            raise ValkeyCommandError("HSET", exc) from exc

    def get_policy(self, tool_name: str) -> ToolPolicy | None:
        return self._policies.get(tool_name)

    def list_policy_names(self) -> list[str]:
        return list(self._policies.keys())

    async def invalidate_by_tool(self, tool_name: str) -> int:
        with self._telemetry.tracer.start_as_current_span(
            "agent_cache.tool.invalidate_by_tool"
        ) as span:
            try:
                span.set_attribute("cache.tool_name", tool_name)
                pattern = (
                    f"{escape_glob_pattern(self._name)}"
                    f":tool:{escape_glob_pattern(tool_name)}:*"
                )
                deleted_count = 0

                async def on_keys(keys: list[str], client: Any) -> None:
                    nonlocal deleted_count
                    pipe = client.pipeline(transaction=False)
                    for k in keys:
                        pipe.delete(k)
                    try:
                        results = await pipe.execute()
                    except Exception as exc:
                        raise ValkeyCommandError("DEL", exc) from exc
                    for count in results:
                        deleted_count += count or 0

                await cluster_scan(self._client, pattern, on_keys)
                span.set_attribute("cache.deleted_count", deleted_count)
                return deleted_count

            except Exception as exc:
                span.record_exception(exc)
                raise

    async def invalidate(self, tool_name: str, args: Any) -> bool:
        hash_ = tool_cache_hash(args)
        key = self._build_key(tool_name, hash_)
        try:
            deleted = await self._client.delete(key)
            return deleted > 0
        except Exception as exc:
            raise ValkeyCommandError("DEL", exc) from exc

    async def refresh_policies(self) -> bool:
        """Refresh policies from Valkey with an atomic swap.

        Clears the in-memory map and repopulates it from HGETALL so that
        policies deleted externally (HDEL) are also removed locally.

        Returns ``True`` on a successful HGETALL, ``False`` if the call threw.
        Used by the periodic refresh loop to drive the
        ``config_refresh_failed`` counter.
        """
        try:
            raw = await self._client.hgetall(self._policies_key)
        except Exception:
            return False

        next_policies: dict[str, ToolPolicy] = {}
        if raw:
            for tool_name, policy_json in raw.items():
                tool_name_str = (
                    tool_name.decode() if isinstance(tool_name, bytes) else tool_name
                )
                policy_str = (
                    policy_json.decode() if isinstance(policy_json, bytes) else policy_json
                )
                try:
                    data = json.loads(policy_str)
                    next_policies[tool_name_str] = ToolPolicy(ttl=data["ttl"])
                except (json.JSONDecodeError, KeyError):
                    pass  # Skip corrupt entries

        self._policies.clear()
        self._policies.update(next_policies)
        return True

    async def load_policies(self) -> None:
        """Load policies from Valkey. Delegates to refresh_policies()."""
        await self.refresh_policies()

    def reset_policies(self) -> None:
        """Clear in-memory policies. Called by AgentCache.flush()."""
        self._policies.clear()

    async def _set(self, key: str, value: str, ttl: int | None) -> None:
        try:
            if ttl is not None:
                await self._client.set(key, value, ex=ttl)
            else:
                await self._client.set(key, value)
        except Exception as exc:
            raise ValkeyCommandError("SET", exc) from exc

    async def _inc_stats(self, updates: dict[str, int]) -> None:
        try:
            pipe = self._client.pipeline(transaction=False)
            for field, amount in updates.items():
                pipe.hincrby(self._stats_key, field, amount)
            await pipe.execute()
        except Exception:
            pass
